### PR TITLE
#11720: Enable ttnn binary to be built using rpath as origin

### PR DESCRIPTION
**Pull request by tapspatel**
This PR enables ttnn to dynamically link to other libraries starting from the location of where the binary is located. This is very useful for upstream projects that want to build ttnn.so and ship it around. It will also help metal's packaging and releases as it allows the package to be relocatable. 

Hierarchy of linker paths
1. directories specified by link flags (-l)
2. directories in rpath and runpath
3. env variables (LD_LIBRARY_PATH)
4. default system lib paths

built in absolute paths are bypassed by this search

### CODE REVIEW

**Comment by tapspatel** [comment]
fyi @ayerofieiev-tt

**Review by ayerofieiev-tt — APPROVED** [review]
Hopefully this does not break existing packaging

**Comment by ayerofieiev-tt** [comment]
@nsmithtt , yes CI make sure to run PostCommit pipeline

**Comment by tapspatel** [comment]
CI pipeline: https://github.com/tenstorrent/tt-metal/actions/runs/10494483049

**Comment by rayraykay** [comment]
@tt-rkim is my tenstorrent account, that's my personal account @nsmithtt . Thanks for outing me :)

Anyway, yes, Artem is right. Although, nothing is stopping you from placing the SOs and headers into the findable include and dynamic lib paths (`/usr/lib` or whatever) yourself.

We hope to add this functionality as part of better distribution work for our software. @TT-billteng can speak more to this

**Review by tt-rkim — APPROVED** [review]

**Comment by tt-rkim** [comment]
I asked this privately to @tapspatel , however could you guys please explain the specific error you're seeing, and where you're building and placing the metal SOs?

**Comment by nsmithtt** [comment]
> I asked this privately to @tapspatel , however could you guys please explain the specific error you're seeing, and where you're building and placing the metal SOs?

We have separate runtime interface that links against ttnn and ttmetal so we need to include those binaries as part of our package.  Currently we are just linking with absolute paths to these libs which is fine for local dev.  But we want to setup our CI to build on GitHub runners, which produce a "release" wheel that then gets installed on the silicon runners in a separate pipeline stage, with no source/build artifacts from the prev stage, and then run all silicon tests.

By adding ORIGIN to RPATH we're able to tell the loader to look for linked dylibs in a path relative to the current dylib, i.e. it allows us to copy _ttnn.so anywhere and as long as libtt_metal.so is next to it, the loader will find libtt_metal.so.  This majorly simplifies packaging into a python wheel, because then the wheel could be installed in some venv w/ any path and it'll work.

There is a caveat that we're working on, which is that the device runtime sources do need to be present as well, so we're trying to see if we can package those in a relocatable way too.

**Comment by nsmithtt** [comment]
I think currently metal seems to be calling patchelf on the fly to add this rpath, for example a system without patchelf fails with:
```
>>> import ttnn
/bin/sh: 1: patchelf: not found
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/nsmith/tmp2/env/lib/python3.8/site-packages/ttnn/__init__.py", line 16, in <module>
    import tt_lib as _tt_lib
  File "/home/nsmith/tmp2/env/lib/python3.8/site-packages/tt_lib/__init__.py", line 77, in <module>
    _check_so_rpath_in_build_lib()
  File "/home/nsmith/tmp2/env/lib/python3.8/site-packages/tt_lib/__init__.py", line 53, in _check_so_rpath_in_build_lib
    subprocess.check_call(f"patchelf --set-rpath {new_rpath} {eager_so}", shell=True)
  File "/usr/lib/python3.8/subprocess.py", line 364, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command 'patchelf --set-rpath /home/nsmith/tmp2/env/lib/python3.8/site-packages/tt_lib/build/lib /home/nsmith/tmp2/env/lib/python3.8/site-packages/tt_lib/build/lib/libtt_eager.so' returned non-zero exit status 127.
```

This change would enable you to not have to run patchelf.

**Comment by tt-rkim** [comment]
> This change would enable you to not have to run patchelf or need it installed.

That's true if `libtt_metal.so` is in the same dir, which in your case sounds like it is. In our wheel it's not, but that's not a relevant problem.

OK - sounds good. I'm wondering if we should make a P3 issue to get rid of this once we get proper distribution working so downstream users (ie. mlir) can use `ttnn` without it choking because it can't find `tt_metal.so`.

**Comment by nsmithtt** [comment]
> > This change would enable you to not have to run patchelf or need it installed.
> 
> That's true if `libtt_metal.so` is in the same dir, which in your case sounds like it is. In our wheel it's not, but that's not a relevant problem.
> 
> OK - sounds good. I'm wondering if we should make a P3 issue to get rid of this once we get proper distribution working so downstream users (ie. mlir) can use `ttnn` without it choking because it can't find `tt_metal.so`.

I'd recommend putting them all in the same dir, it'll make your lives easier and it's a bit unconventional for a single wheel to install multiple site-packages.  But RPATH is programmable during compile time so if you wanted to you could add `"$ORIGIN/tt_metal:$ORIGIN/tt_eager:$ORIGIN/tt_lib"`.

**Comment by nsmithtt** [comment]
> We're about to do some work on the wheel, and honestly that's not a bad idea. We may go with that.
> 
> In any case, it doesn't sound like MLIR is consuming as wheel, so I'm assuming that work is not blocking you guys, and you guys should be able to move forward when this change makes it in.

Yeah, not blocking! This PR is all we need

**Comment by TT-billteng** [comment]
We're planning to use to use the wheel + docker as the default method to run any python post-commit jobs

**Comment by tapspatel** [comment]
Main post-commit is currently red: https://github.com/tenstorrent/tt-metal/actions/runs/10676173099 Waiting for green pipelines before I merge

**Comment by tapspatel** [comment]
Kicked off another pipeline: https://github.com/tenstorrent/tt-metal/actions/runs/10693095366

**Comment by tapspatel** [comment]
All tests passed: https://github.com/tenstorrent/tt-metal/actions/runs/10693095366 merging.

**Pull request merged**
